portable/NetworkInterface/TM4C/NetworkInterface.c

```diff
@@ -0,0 +1,599 @@
+        _rx_descriptors[i].ui32CtrlStatus = DES0_RX_CTRL_OWN;
+
+        // Set the next link the DMA descriptor chain
+        _rx_descriptors[i].DES3.pLink = ((i == (SYSCONFIG_FREERTOS_TCP_RX_DESC_COUNT - 1)) ? &_rx_descriptors[0] : &_rx_descriptors[i + 1]);
+    }
+
+    // Set the RX descriptor index
+    _rx_descriptor_list.write = 0;
+
+    // Set the head of the DMA descriptor list in the EMAC peripheral
+    MAP_EMACTxDMADescriptorListSet(EMAC0_BASE, &_tx_descriptors[0]);
+    MAP_EMACRxDMADescriptorListSet(EMAC0_BASE, &_rx_descriptors[0]);
+}
+
+void freertos_tcp_ethernet_int(void)
+{
+    uint32_t status;
+
+    // Read the interrupt status
+    status = EMACIntStatus(EMAC0_BASE, true);
+
+    // Handle power management interrupts
+    if(status & EMAC_INT_POWER_MGMNT)
+    {
+        MAP_EMACTxEnable(EMAC0_BASE);
+        MAP_EMACRxEnable(EMAC0_BASE);
+
+        MAP_EMACPowerManagementStatusGet(EMAC0_BASE);
+
+        status &= ~(EMAC_INT_POWER_MGMNT);
+    }
+
+    if (status)
+    {
+        MAP_EMACIntClear(EMAC0_BASE, status);
+    }
+
+    // Handle PHY interrupts
+    if (EMAC_INT_PHY & status)
+    {
+        _process_phy_interrupt();
+    }
+
+    // Handle Transmit Complete interrupts
+    if (EMAC_INT_TRANSMIT & status)
+    {
+        _process_transmit_complete();
+    }
+
+    // Handle Receive interrupts
+    if ((EMAC_INT_RECEIVE | EMAC_INT_RX_NO_BUFFER | EMAC_INT_RX_STOPPED) & status)
+    {
+        _process_received_packet();
+    }
+}
+
+static void _process_transmit_complete(void)
+{
+    uint32_t i;
+    tEMACDMADescriptor *dma_descriptor;
+    NetworkBufferDescriptor_t *stack_descriptor;
+
+    for (i = 0; ((i < _tx_descriptor_list.number_descriptors) && (_tx_descriptor_list.read != _tx_descriptor_list.write)); i++)
+    {
+        // Get a reference to the current DMA descriptor
+        dma_descriptor = &_tx_descriptors[_tx_descriptor_list.read];
+
+        // If the descriptor is still owned by the DMA controller, exit
+        if (dma_descriptor->ui32CtrlStatus & DES0_TX_CTRL_OWN)
+        {
+            break;
+        }
+
+        // Get the 'hidden' reference to the stack descriptor from the buffer
+        stack_descriptor = *((NetworkBufferDescriptor_t **) (((uint8_t *) dma_descriptor->pvBuffer1) - ipBUFFER_PADDING));
+
+        // Release the stack descriptor
+        vNetworkBufferReleaseFromISR(stack_descriptor);
+
+        _tx_descriptor_list.read++;
+
+        if (_tx_descriptor_list.read == _tx_descriptor_list.number_descriptors)
+        {
+            _tx_descriptor_list.read = 0;
+        }
+    }
+}
+
+static void _process_received_packet(void)
+{
+    NetworkBufferDescriptor_t *new_stack_descriptor;
+    NetworkBufferDescriptor_t *cur_stack_descriptor;
+    tEMACDMADescriptor *dma_descriptor;
+    BaseType_t higher_priority_task_woken = pdFALSE;
+    uint32_t i;
+
+    // Go through the list of RX DMA descriptors
+    for (i = 0; i < SYSCONFIG_FREERTOS_TCP_RX_DESC_COUNT; i++)
```

Comment:
This ISR does a lot of processing, the code will call several `FromISR` functions. It looks correct to me, but personally I would find it more elegant to just wakeup the MAC-task `_packet_received_thread()`, and let that function work on the DMA queues.
[Here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/55b3b94eb8b4cfae320d20386e8c3f5b1dea13be/portable/NetworkInterface/STM32Fxx/NetworkInterface.c#L316) is an example of how a network ISR will delegate the actual processing to the EMAC thread. This is done for both reception, and for checking the transmission buffers.

It should not be complicated to move the code from the ISR to the deferred interrupt handler.
You will not need the queue `_received_packets_queue` anymore.
The function can block by calling :
~~~
ulTaskNotifyTake( pdFALSE, ulMaxBlockTime );
~~~
Remember that it should also [check the Link Status](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/55b3b94eb8b4cfae320d20386e8c3f5b1dea13be/portable/NetworkInterface/STM32Fxx/NetworkInterface.c#L1301) from time to time.
As long as the Link Status is low, the driver should not attempt to send packets, because that may block and fill all network buffers.

Reply:
I see what you're saying here. From the porting guide, it seemed to be implied that it was more advantageous to handle RX packets in the ISR. I'll take a look.